image.py

```diff
@@ -0,0 +1,355 @@
+
+    -   Capable of building docker images from dockerfiles or dockerfile-
+        formatted strings via the build classmethods.
+    -   Capable of acquiring information about the given image (see property
+        methods below).
+    -   Capable of running commands on containers built from the image using
+        the run method.
+    -   All docker build and run methods on the class can recive a file-like
+        object, (io.TextIOBase or some subclass of it,) and they will write the
+        outputs from docker into the file to enable logging.
+
+    NOTE: This class does not work correctly if the image built does not have
+        bash installed.
+    """
+
+    def __init__(self, name_or_id: str):
+        """Initialize this Image. Connects the class with its ID.
+
+        Parameters
+        ----------
+        name_or_id : str
+            A name or ID by which to find this image using docker inspect.
+        """
+        self._id = get_image_id(name_or_id)
+
+    @classmethod
+    def build_from_dockerfile(
+        cls,
+        context: str,
+        tag: str,
+        output_file: Union[io.TextIOBase, None] = None,
+        dockerfile_loc: str = ""
+    ):
+        """Build a Dockerfile at the given path with the given name, then
+        return the associated Image instance.
+
+        Parameters
+        ----------
+        context : str
+            The path to the file containing the Dockerfile
+        tag : str
+            A name for the image
+        output_file : io.TextIOBase or None
+            A file-like object that the sterr output of the docker build
+            program will be written to. If None, the method will not output to
+            a file. Defaults to None.
+        dockerfile_loc : str
+            The location of the Dockerfile to build, relative to the position
+            held at the context argument described above. If empty, the command
+            is called assuming that the dockerfile is held at the context root
+            folder. Defaults to "".
+
+        Returns
+        -------
+        Image
+            An Image class instance that references a Docker image built from
+            the Dockerfile at context/dockerfile_loc
+
+        Raises
+        -------
+        CalledProcessError
+            Raised if the docker build command fails. Holds the value returned
+            by the docker build command in its returncode attribute, if not 0.
+        """
+
+        if dockerfile_loc:
+            dockerfile_arg = f" --file='{dockerfile_loc}'"
+        else:
+            dockerfile_arg = ""
+
+        command = "docker build --network=host " + \
+            f"{context}{dockerfile_arg} -t {tag}"
+        process = run(
+            split(command),
+            check=True,
+            text=True,
+            capture_output=True)
+
+        if output_file is not None and isinstance(output_file, io.TextIOBase):
+            output_file.write(process.stderr)
```

Comment:
With using `popen()`, there may be some additional complexities with building from strings since `run()` provides a pretty easy `input` argument whereas I believe `communicate()` has to be used to talk to `popen` subprocesses.

Reply:
Still looking at this. It's hard to find documentation on this that isn't 7+ years old.

Reply:
I decided this issue probably wasn't as hard as I was making it out to be, so I took a stab at it. [(commit)](https://github.com/isce3-testing/image-sandbox/pull/1/commits/3e3cf7c310d16da439e18f64dcc8deaaaed49a29)

I don't think `subprocess.run()` is compatible with this functionality so I reverted to `subprocess.Popen` and streamed the output in a loop. This method can capture output to print, write to file, and return simultaneously and in real time. It was also simple enough to write it to its own function (`_run_process()`, which may be better outside of the Image class than inside of it) to avoid re-writing repetitive code since this method is a little more complicated. It also opens up the door to allowing things like passing inputs into `run()` without having to refactor much code, if that ever becomes necessary.

Reply:
Hmm, hand-rolling our own `_run_process()` function using the lower-level `subprocess.Popen` interface involves a lot more complexity and potential pitfalls -- and I'm still convinced that what we want to achieve can be done simply with `subprocess.run()`.

Here's a minimal example using `subprocess.run()` that demonstrates the expected behavior in my testing:

```bash
# ~ emit-messages.sh ~

# Enable common error handling options.
set -o errexit
set -o nounset
set -o pipefail

# Print interleaved messages to stdout & stderr.
echo "INFO MESSAGE 1"
echo "ERROR MESSAGE 1" >&2
echo "INFO MESSAGE 2"
echo "ERROR MESSAGE 2" >&2
```

```python
# ~ run_stdout_stderr.py ~

import subprocess
from pathlib import Path

if __name__ == "__main__":
    # Run a subprocess and stream both stdout & stderr to the same text file.
    with Path("logfile.txt").open("w") as logfile:
        subprocess.run(
            ["bash", "emit-messages.sh"],
            stdout=logfile,
            stderr=logfile,
            check=True,
        )
```

```console
$ python run_stdout_stderr.py
$ cat logfile.txt
INFO MESSAGE 1
ERROR MESSAGE 1
INFO MESSAGE 2
ERROR MESSAGE 2
```

The downside of this approach is that it can't be used together with the `capture_output=True` option of `subprocess.run()`. So we might need two flavors of `Image.run()` -- one that captures and returns its output, and one that can stream its output & error messages to files (or to the console). Still I think this probably better than going the `Popen` route.

Reply:
Modified to fit the above thoughts per our discussion earlier today. Type checking is currently a little weird due to interfacing directly with the subprocess.run() method (namely using PIPE and STDOUT as arguments, which are themselves ints)

also, the use of run() by both Image and subprocess is feeling like it might confuse others, I'm thinking about a good rename for Image.run()